bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs

```diff
@@ -0,0 +1,781 @@
+		InboundQueueLocation,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	)>,
+}
+
+impl<
+		CreateAssetCall,
+		CreateAssetDeposit,
+		EthereumNetwork,
+		InboundQueueLocation,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	>
+	MessageToXcm<
+		CreateAssetCall,
+		CreateAssetDeposit,
+		EthereumNetwork,
+		InboundQueueLocation,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	>
+where
+	CreateAssetCall: Get<CallIndex>,
+	CreateAssetDeposit: Get<u128>,
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueueLocation: Get<InteriorLocation>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	/// Parse the message into an intermediate form, with all fields decoded
+	/// and prepared.
+	fn prepare(message: Message) -> Result<PreparedMessage, ConvertMessageError> {
+		let ether_location = Location::new(2, [GlobalConsensus(EthereumNetwork::get())]);
+
+		let remote_xcm: Xcm<()> = match &message.xcm {
+			XcmPayload::Raw(raw) => Self::decode_raw_xcm(raw),
+			XcmPayload::CreateAsset { token, network } =>
+				Self::make_create_asset_xcm(token, *network, message.value)?,
+		};
+
+		// Asset to cover XCM execution fee
+		let execution_fee_asset: Asset = (ether_location.clone(), message.execution_fee).into();
+
+		let mut assets = vec![];
+
+		let claimer: Option<Location> = match message.claimer {
+			Some(claimer_bytes) => Location::decode(&mut claimer_bytes.as_ref()).ok(),
+			None => None,
+		};
+
+		if message.value > 0 {
+			// Asset for remaining ether
+			let remaining_ether_asset: Asset = (ether_location.clone(), message.value).into();
+			assets.push(AssetTransfer::ReserveDeposit(remaining_ether_asset));
+		}
+
+		for asset in &message.assets {
+			match asset {
+				EthereumAsset::NativeTokenERC20 { token_id, value } => {
+					let token_location: Location = Location::new(
+						2,
+						[
+							GlobalConsensus(EthereumNetwork::get()),
+							AccountKey20 { network: None, key: (*token_id).into() },
+						],
+					);
+					let asset: Asset = (token_location, *value).into();
+					assets.push(AssetTransfer::ReserveDeposit(asset));
+				},
+				EthereumAsset::ForeignTokenERC20 { token_id, value } => {
+					let asset_loc = ConvertAssetId::convert(&token_id)
+						.ok_or(ConvertMessageError::InvalidAsset)?;
+					let mut reanchored_asset_loc = asset_loc.clone();
+					reanchored_asset_loc
+						.reanchor(&GlobalAssetHubLocation::get(), &EthereumUniversalLocation::get())
+						.map_err(|_| ConvertMessageError::CannotReanchor)?;
+					let asset: Asset = (reanchored_asset_loc, *value).into();
+					assets.push(AssetTransfer::ReserveWithdraw(asset));
+				},
+			}
+		}
+
+		let topic = blake2_256(&("SnowbridgeInboundQueueV2", message.nonce).encode());
```

Comment:
I'd suggest the topic id generated from the source chain, it's important for building a global indexer.

Btw:  Except for XCM on AH, the topic should also be inserted into the XCM instructions on the destination chain.

Reply:
Another option as we discussed is to let UI to generate topic directly and insert into the xcm instructions on the destination chain.

Reply:
We should check the last instruction of the xcm, and if it is not SetTopic, we should append SetTopic to the instructions based on this hash. Look at the `WithUniqueTopic` as an example.
https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/routing.rs/#L47-L53
This will allow the UI to generate a custom topic should they choose, but if not passed in by the UI, a deterministic topic will can be generated from the nonce.

For examples where there are multi-hop sends, the UX will set the same topic for each hop when building xcm. But for simple transfers to asset hub, the UX could leave it out.

Reply:
Done in https://github.com/paritytech/polkadot-sdk/pull/7402/commits/34631cfb9c8d13c6c5f6bbd123124075e3b20904.